### ISSUE TITLE
Excessively long time spent in retraining model

### ISSUE THREAD

**Issue by VictorBrouwers** [mir-group/flare#296]
**Describe the bug**
In the described problem, I'm recreating the original FLARE paper result of melting Al. 
I'm not really facing a bug, but rather a significant time bottleneck that seems not to be there in the published FLARE results: after about 20-30 DFT steps, the time required to create the (parallel) Ky_matrix (in gp_algebra.py) and retrain the GP (in gp.py) starts to grow at a very large pace and reaches an order of 3000 - 10.000 s per step after a single DFT calculation (the DFT calculation duration itself remains about constant), compared to a maximum of around 800 s for the original FLARE paper. Running the script with cProfile confirms these points are indeed the bottleneck. 
It appears that for the parallel K_matrix creation and retraining step, only 1 core is used on my machines, while during the 'regular MD' steps, multiple cores are used. I'm wondering if this behavior is intended or not and if this could be part of my problem.
I'm running on 32 cores (tried more, but doesn't change the results) with sufficient memory.

**To Reproduce**
Not sure if possible. I have installed FLARE both on a personal workstation and on Snellius, a Dutch national supercomputer, and I get similar behavior on both systems. 
The Snellius system has 'mir-flare 0.1.7' pip installed, 'GCCcore/10.3.0' and ' Python/3.9.5-GCCcore-10.3.0'.

**Expected behavior**
My simulation is simply that of melting Al, similar to the first FLARE paper. From that paper, I've checked the results, in particular the time required per calculation step, which looks like this:
![image](https://user-images.githubusercontent.com/66886615/145002148-b74c74dd-06be-4eec-9471-43d7f5690c58.png).

These calculation times would be reasonable for me. Note that this simulation is 10 ps with a temperature kick at 5 ps.

**Screenshots**
In my simulation, I only run for about 200 fs, and give the temperature kick at 100 fs. The amount of DFT calculation steps I need is similar / less than the 10 ps version in the FLARE paper. However, the time required per step looks more like this for me:
![image](https://user-images.githubusercontent.com/66886615/145002570-c66d3729-27f8-4604-a74b-130b4ae34e8d.png)

In more extreme cases, but still at only around 100 DFT and retraining steps, the behaviour looks like this:
![image](https://user-images.githubusercontent.com/66886615/145002802-42ea96f9-5f44-4a54-af5c-c459f74fc1bc.png).

**Additional context**
I've included my script and otf.out (of an incomplete run, 136/200 time steps) for verification, see [Github_issue.zip](https://github.com/mir-group/flare/files/7667363/Github_issue.zip).
Furthermore, it seems that 'regular MD' steps are also significantly slower than the published FLARE results. This may indicate that my parallelization is not functional. However, my output does indicate 32 cores, and during the 'regular MD' steps, I do see 32 cores fire up in my machine. 

Would love to hear from you, and please let me know if I can provide more information to help out.

Kind regards,
Victor Brouwers

**Comment by jonpvandermause** [comment]
Hi Victor,

Thanks for the detailed issue report. One thing that immediately jumps to mind: in the current version of the code, the default setting is to train on potential energies. This is extremely expensive with exact GPs, because it requires you to store all environments in the structure and loop over them when you predict/train. In the paper, we train on force components only (and this is what I would recommend with the exact 2+3 GPs in this repo).

Can you try adding `force_only=True` to your `otf_params` dictionary? This will ensure that only force labels get added to the training set during OTF.

I'm doing some tests with your script now to see if this helps on our cluster.

Best,
Jon

**Comment by VictorBrouwers** [comment]
Hi Jon, 

Thanks for the quick reply. This has solved the mystery for me! 
My current calculation times look like the image enclosed below, so very much in line with your published results.
![Calc_D_time](https://user-images.githubusercontent.com/66886615/145092223-9ee60eab-d7ec-4e06-9c75-e21eabc5b5ea.png)

Kind regards,
Victor Brouwers

**Comment by jonpvandermause** [comment]
Hi Victor,

That's great, I'm glad to hear it's working now! I think we'll probably make "force only" training the default to avoid this issue in the future.

Best,
Jon
